apps/emqx_mq/src/emqx_mq_message_db.erl

```diff
@@ -0,0 +1,314 @@
+            Topic = mq_message_topic(MQ, ClientId),
+            Value = encode_message(Message),
+            maps:update_with(Shard, fun(Msgs) -> [{Topic, Value} | Msgs] end, [{Topic, Value}], Acc)
+        end,
+        #{},
+        Messages
+    ),
+    grouped_by_shard_to_list(ByShard);
+group_by_shard(#{is_lastvalue := true, topic_filter := _TopicFilter} = MQ, Messages) ->
+    ByShard = lists:foldl(
+        fun(Message, Acc) ->
+            Props = emqx_message:get_header(properties, Message, #{}),
+            UserProperties = maps:get('User-Property', Props, []),
+            Key = proplists:get_value(
+                ?MQ_KEY_USER_PROPERTY, UserProperties, undefined
+            ),
+            case Key of
+                undefined ->
+                    ?tp(warning, mq_message_db_insert_error, #{
+                        mq => _TopicFilter,
+                        from => emqx_message:from(Message),
+                        reason => undefined_key
+                    }),
+                    Acc;
+                _ ->
+                    Shard = emqx_ds:shard_of(?MQ_MESSAGE_LASTVALUE_DB, Key),
+                    Topic = mq_message_topic(MQ, Key),
+                    Value = encode_message(Message),
+                    maps:update_with(
+                        Shard, fun(Msgs) -> [{Topic, Value} | Msgs] end, [{Topic, Value}], Acc
+                    )
+            end
+        end,
+        #{},
+        Messages
+    ),
+    grouped_by_shard_to_list(ByShard).
+
+grouped_by_shard_to_list(ByShard) ->
+    lists:map(
+        fun({Shard, TVs}) ->
+            {Shard, lists:reverse(TVs)}
+        end,
+        maps:to_list(ByShard)
+    ).
+
+format_insert_tx_results(Results) ->
+    format_insert_tx_results(Results, []).
+
+format_insert_tx_results([] = _Results, [] = _ErrorAcc) ->
+    ok;
+format_insert_tx_results([] = _Results, ErrorAcc) ->
+    {error, lists:reverse(ErrorAcc)};
+format_insert_tx_results([{atomic, _Serial, ok} | Results], ErrorAcc) ->
+    format_insert_tx_results(Results, ErrorAcc);
+format_insert_tx_results([{error, _, _} = Error | Results], ErrorAcc) ->
+    format_insert_tx_results(Results, [Error | ErrorAcc]).
+
+delete(DB, Topic) ->
+    lists:foreach(
+        fun({Shard, Generation}) ->
+            TxOpts = #{
+                db => DB,
+                shard => Shard,
+                generation => Generation,
+                sync => true,
+                retries => ?MQ_MESSAGE_DB_APPEND_RETRY
+            },
+            emqx_ds:trans(TxOpts, fun() ->
+                emqx_ds:tx_del_topic(Topic)
+            end)
+        end,
+        maps:keys(emqx_ds:list_slabs(DB))
+    ).
+
+mq_message_topic(#{topic_filter := TopicFilter, id := Id} = _MQ, Key) ->
+    ?MQ_MESSAGE_DB_TOPIC(TopicFilter, Id, Key).
+
+db(#{is_lastvalue := true} = _MQ) ->
+    ?MQ_MESSAGE_LASTVALUE_DB;
+db(#{is_lastvalue := false} = _MQ) ->
+    ?MQ_MESSAGE_REGULAR_DB.
+
+insert_generation(?MQ_MESSAGE_LASTVALUE_DB) ->
+    1;
```

Comment:
Q: shouldn't this be dynamic (i.e., depend on the current generation value)?

Reply:
No, last-value queues do not switch generations, because their data is removed with topic del opts.